### PR TITLE
CRC Check + Open Addressing Hash Map

### PR DESCRIPTION
**Pull request by Maithem**
## Overview

Improve the metadata overhead per write.

Check Checksum on cache misses.

Why should this be merged: Fixes a bug (CRC checks on cache misses) and reduces the log unit memory usage. 

Related issue(s) (if applicable): #<number>

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/SegmentHandle.java [hunk 1]
```diff
@@ -34,9 +35,8 @@
     @NonNull
     String fileName;
 
-    private final Map<Long, AddressMetaData> knownAddresses = new ConcurrentHashMap<>();
-    private final Set<Long> trimmedAddresses = Collections.newSetFromMap(new ConcurrentHashMap<>());
-    private final Set<Long> pendingTrims = Collections.newSetFromMap(new ConcurrentHashMap<>());
+    private final Map<Long, Long> knownAddresses = new Long2LongHashMap(-1);
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/SegmentHandle.java]: Set table capacity to shorten the linear probe.

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/SegmentHandle.java [hunk 2]
```diff
@@ -34,9 +35,10 @@
     @NonNull
     String fileName;
 
-    private final Map<Long, AddressMetaData> knownAddresses = new ConcurrentHashMap<>();
-    private final Set<Long> trimmedAddresses = Collections.newSetFromMap(new ConcurrentHashMap<>());
-    private final Set<Long> pendingTrims = Collections.newSetFromMap(new ConcurrentHashMap<>());
+    private final Map<Long, Long> knownAddresses = new Long2LongHashMap(StreamLogFiles.RECORDS_PER_LOG_FILE,
+            0.67F, // Load factor
+            -1); // missing value
+
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/SegmentHandle.java]: This map can be represented with only an array.

Comment (Maithem) [reply infrastructure/src/main/java/org/corfudb/infrastructure/log/SegmentHandle.java]: use long[] index = new long[RECORDS_PER_LOG_FILE] instead.

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java [hunk 3]
```diff
@@ -693,15 +695,25 @@ private void readAddressSpace(SegmentHandle segment) throws IOException {
...
+        if (!segment.getKnownAddresses().containsKey(address)) {
             return null;
         }
 
         try {
-            ByteBuffer entryBuf = ByteBuffer.allocate(metaData.length);
-            fileChannel.read(entryBuf, metaData.offset);
+            long metaDataOffSet = segment.getKnownAddresses().get(address);
+            // TODO(Maithem): Pre-allocate a large buffer to avoid the second read
+            ByteBuffer metaData = ByteBuffer.allocate(METADATA_SIZE);
```
Comment (xiaoqin2012) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java]: Please change metaData here to metaDataBuffer. As the same variable name is used again for MetaData class object two lines below.

Comment (Maithem) [reply infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java:705]: ok

**Comment by Maithem** [comment]
Moving to a an embedded where the index is stored on disk is a cleaner solution than this PR's optimization.

**Pull request closed without merging**